codex/validation.nim

```diff
@@ -43,13 +46,26 @@ proc waitUntilNextPeriod(validation: Validation) {.async.} =
   trace "Waiting until next period", currentPeriod = period
   await validation.clock.waitUntil(periodEnd.truncate(int64) + 1)
 
+func partitionIndexForSlotId(validation: Validation, slotId: SlotId): int =
+  if (validation.validationParams.partitionSize == 1):
+    0
+  else:
+    let slotIdUInt256 = UInt256.fromBytesBE(slotId.toArray)
```

Comment:
Yes, I believe it is good. EVM is big endian and here I am explicitly using `fromBytesBE`. The `fromBytes` uses the native (i.e. machine) endianness, which indeed would not be portable. By the way, `stint` v2 changes the default behaviour - it defaults to BE instead of native (machine) endianness, but we are still using the old version of `stint` package. `fromBytesBE` seems to be then a right option.

Reply:
Sure, but the slotid isn't generated in the EVM if I'm not mistaken? All-in-all it's probably ok, but we may need to ensure that converting to an array using the right endianness. I could be wrong in that assumption as well.

Reply:
I think it is done both in the client and in the contracts using the same kind of approach.

On-chain (https://github.com/codex-storage/codex-contracts-eth/blob/bba8736132007e52f560dd8f49e4a9b89b03eb6e/contracts/Requests.sol#L52):

```solidity
function slotId(
    RequestId requestId,
    uint256 slotIndex
  ) internal pure returns (SlotId) {
    return SlotId.wrap(keccak256(abi.encode(requestId, slotIndex)));
  }
```

In the client (https://github.com/codex-storage/nim-codex/blob/1e2ad956591d41ec93ede1dec7da228a4558e9d3/codex/contracts/requests.nim#L166):

```nim
func slotId*(requestId: RequestId, slot: UInt256): SlotId =
  let encoding = AbiEncoder.encode((requestId, slot))
  SlotId(keccak256.digest(encoding).data)
```

Because it is a digest in the end, the endianness should not matter that match as long as it is kept consistent on each architecture, right? I checked initially with Mark, he leans towards big endian as well.

Reply:
I likely was not clear, my apologies. I wasn't trying to make a point that we should or shouldn't use BE, my point is that we should be consistent in *both conversion directions* since endianness is not portable. In this particular case, I was asking, "are you sure BE is the right solution" purely because it's not clear to me that we are using BE when converting from `SlotId` to `array[32, byte]`, which happens here: https://github.com/codex-storage/nim-codex/blob/4ab55c63c25d6dca29dfaea29487ce3ed23bd039/codex/contracts/requests.nim#L60-L61

Reply:
Apologies in advance if I am not getting it right - it is a lot of internals to parse :), but I tried to drill a bit...

The function you mentioned, is just getting around the fact that `SlotId` is already `array[32, byte]` but it is `distinct`: `SlotId* = distinct array[32, byte]` (same applies to `RequestId`) The function you probably want to be concerned with is this:

```nim
func slotId*(requestId: RequestId, slot: UInt256): SlotId =
  let encoding = AbiEncoder.encode((requestId, slot))
  SlotId(keccak256.digest(encoding).data)
```

and to be precise, the key seems to be here how the slot index - `slot: UInt256` above gets encoded when getting JSONs from RPC interface. It might be this one (`vendor/nim-ethers/ethers/providers/jsonrpc/conversions.nim`):

```nim
func fromJson*(json: JsonNode, name: string, result: var UInt256) =
  result = UInt256.fromHex(json.getStr())
```

And so, it may all get down to `UInt256.fromHex` in `vendor/stint/stint/io.nim`:

```nim
func fromHex*(T: typedesc[StUint|StInt], s: string): T {.inline.} =
  ## Convert an hex string to the corresponding unsigned integer
  parse(s, type result, radix = 16)
```

`parse` follows the input hex string, and convert each encountered byte to `Uint256` by doing: 

```nim
input[curr].readHexChar.stuint(bits)
```

we have `func readHexChar(c: char): int8 {.inline.}=` and thus the final conversion happens here:

```nim
func stuint*[T: SomeInteger](n: T, bits: static[int]): StUint[bits] {.inline.}=
```

So, it looks that the json hex will be encoded into UInt256 that follows the underlying architecture when it comes to endianness:

```nim
type
  # ### Private ### #
  UintImpl*[BaseUint] = object
    when system.cpuEndian == littleEndian:
      lo*, hi*: BaseUint
    else:
      hi*, lo*: BaseUint
```

It looks therefore, that the the result of `func slotId*(requestId: RequestId, slot: UInt256): SlotId =` is consistent across architectures (otherwise we would probably already be in trouble).

With all this in mind, I now start thinking that we should use the default version of `fromBytes` to keep the low level implementation correct:

Thus we probably need to use this:

```nim
let slotIdUInt256 = UInt256.fromBytes(slotId.toArray)
```

As you said running some experiment on a big endian machine would be nice thing to do in order to confirm that... otherwise it is quite hard to be sure...

> We may need to pay attention when updating Stint in the future.

Reply:
Yes, you're right, the conversion does start in the `AbiEncoder`, my apologies. It was getting late when I posted the type conversion snippet :)

> we should use the default version of fromBytes to keep the low level implementation correct

Yes, that is exactly what I meant. Because in that way, we are using the underlying UintImpl conversion in both directions, which should handle the endianness for us automatically.

Reply:
Hmm... I did some more digging and run some experiments on a separate repo. If we dig one step further than in ABI Encoder we will find this (`vendor/nim-contract-abi/contractabi/encoding.nim`):

```nim
func encode(encoder: var AbiEncoder, value: SomeUnsignedInt | StUint) =
  encoder.padleft(value.toBytesBE)
```

So, in functions `toBytesBE`/`fromBytesBE`, the endianness indicates the order in the output/input byte array.
If we use just `toBytes`/`fromBytes` it will apply the machine order to the byte arrays. Taken the above ` encoder.padleft(value.toBytesBE)`, we have to use `fromBytesBE` when converting to `StInt`s.

The rule seems to be the following:

`byte arrays must be always in network order, thus BE!`

I think when they were updating the Stint library, they realised that the byte arrays are best kept in network order and that's why there they use BE as default instead of machine endianness when working with byte arrays.

I will show you this later running. Now I need to switch off for a moment ;).

Reply:
Nice job, Marcin! Now we are sure that `toBytesBE` was indeed the right choice 👏

Reply:
I am happy you triggered more attention to it because now I have better picture!